### PR TITLE
Slice Diversity Reg: Gram matrix orthogonality on slice attention

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver physics-attention mechanism assigns each mesh node to slices via learned soft weights. If multiple slices converge to attend to the same spatial regions (slice collapse), the model wastes representational capacity. By penalizing the cosine similarity between different slice profiles using a Gram matrix orthogonality loss, we force each slice to specialize in a distinct spatial region — boundary layer, wake, freestream, suction peak, etc.

This is a pure loss-side change: **zero architecture modification, zero inference cost.** The penalty encourages diverse spatial decomposition during training, which should improve OOD generalization because each distinct physical phenomenon gets its own dedicated slice budget.

Related to multi-head attention diversity regularization (Li et al., 2018; "Improving Multi-Head Attention with Disagreement Regularization"), adapted for Transolver's slice-assignment attention.

**Confidence:** MEDIUM. Loss-side changes have a mixed track record, but the mechanism is orthogonal to all in-flight experiments and targets a structural capacity issue.

### Step 1: Add config flags

```python
# In Config dataclass in train.py:
slice_diversity_reg: bool = False
slice_diversity_weight: float = 0.005
```

### Step 2: Modify Physics_Attention_Irregular_Mesh.forward() to return diversity loss

The slice attention weights are computed in the Physics_Attention block as:
```python
slice_weights = self.slice_softmax(self.slice_linear(x_mid) / self.slice_temp)
# Shape: [B, H, N, slice_num]
```

After this line, compute diversity loss when enabled. Add a helper or inline the calculation:

```python
# Compute diversity loss from slice attention weights
# slice_weights: [B, H, N, S]
div_loss = torch.tensor(0.0, device=slice_weights.device)
if cfg.slice_diversity_reg:
    # Average over heads: [B, N, S]
    w = slice_weights.mean(dim=1)
    # L2-normalize each slice profile over nodes: [B, N, S]
    w_norm = w / (w.norm(dim=1, keepdim=True) + 1e-6)
    # Gram matrix of cosine similarities: [B, S, S]
    G = torch.bmm(w_norm.transpose(1, 2), w_norm)
    # Penalize off-diagonal entries (cosine sim between different slices should be 0)
    eye = torch.eye(G.shape[-1], device=G.device, dtype=G.dtype).unsqueeze(0)
    div_loss = ((G - eye) ** 2).mean()

# Modify the return of the attention block to include div_loss alongside the output
```

Pass `cfg` (or a boolean flag) into `Physics_Attention_Irregular_Mesh` so it can access `slice_diversity_reg`.

### Step 3: Accumulate diversity loss in TransolverIrregularMesh.forward()

In the block loop, accumulate losses:
```python
total_div_loss = 0.0
for block in self.blocks:
    fx, fx_deep, div_loss = block(fx, x, ...)  # unpack div_loss
    total_div_loss = total_div_loss + div_loss
```

After the block loop, add to total loss:
```python
if cfg.slice_diversity_reg:
    loss = loss + cfg.slice_diversity_weight * (total_div_loss / cfg.n_layers)
```

**torch.compile:** `bmm`, `norm`, `eye` are all compile-safe. No dynamic shapes or conditional tensor ops.

**VRAM impact:** Negligible — Gram matrix is [B=4, 96, 96] = 144KB per block.

### Step 4: Diagnostic check (2 epochs)

```bash
cd cfd_tandemfoil && python train.py --agent askeladd --debug \
  --slice_diversity_reg --slice_diversity_weight 0.005 \
  --wandb_name "askeladd/div-debug" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

Verify: (1) No NaN/Inf, (2) div_loss logs cleanly to W&B, (3) torch.compile passes.

### Step 5: Run 2 seeds

```bash
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/slice-diversity-reg-s42" --wandb_group "askeladd/slice-diversity-reg" \
  --slice_diversity_reg --slice_diversity_weight 0.005 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

Seed 73: identical but `--seed 73 --wandb_name "askeladd/slice-diversity-reg-s73"`.

### Step 6: Optional variation (if time permits)

If both seeds finish, try λ=0.01 with seed 42:
```bash
--wandb_name "askeladd/slice-diversity-reg-w01-s42" --wandb_group "askeladd/slice-diversity-reg" \
--slice_diversity_reg --slice_diversity_weight 0.01
```

## Baseline

Current best (PR #2207, TE Coordinate Frame, 2-seed avg):

| Metric | Baseline | Target |
|--------|----------|--------|
| p_in   | 12.490   | < 12.490 |
| p_tan  | **28.521** | **< 28.521** |
| p_oodc | 7.618    | < 7.618 |
| p_re   | 6.411    | < 6.411 |

W&B runs: obn1wfja (seed 42, p_tan=28.641), 52irfwwg (seed 73, p_tan=28.400).

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-te-coord" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Slice Diversity Regularization

### Results (2 seeds)

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | 12.490   | 13.9    | 13.6    | 13.75      | +10.1% worse |
| p_tan  | **28.521** | **30.9** | **29.1** | **30.00** | **+5.2% worse** |
| p_oodc | 7.618    | 8.0     | 8.4     | 8.20       | +7.6% worse |
| p_re   | 6.411    | 6.8     | 7.0     | 6.90       | +7.6% worse |

**W&B runs:** 2e2pvqll (seed 42), gy4k4lpr (seed 73)
**W&B group:** askeladd/slice-diversity-reg

### Implementation

Added Gram matrix orthogonality penalty on slice attention weights:
- After softmax: compute mean over heads → L2-normalize over nodes → cosine similarity Gram matrix
- Penalize off-diagonal entries: `((G - I)^2).mean()` weighted by 0.005
- Applied across all 3 TransolverBlocks, averaged
- `div_loss` logged to W&B every 50 steps — decreases steadily from ~0.8 to ~0.1 (diversity improves)

### Analysis

**Negative across all metrics.** The diversity loss successfully forces slices apart (div_loss drops from 0.8→0.1), but this hurts performance:

1. **Slice collapse may be beneficial:** The current model likely converges multiple slices onto critical regions (boundary layer, wake) because those regions need more representational capacity. Forcing diversity redistributes capacity to less important regions (freestream)
2. **Gradient interference:** The diversity penalty creates opposing gradients against the main task loss — the model wants to concentrate slices on hard regions, but the penalty pushes them apart
3. **Only 134 epochs reached** (vs 148-149 for baseline) — the extra loss term may slow convergence. At 80s/epoch vs 72-73s baseline, the regularization adds some overhead

### Suggested follow-ups
1. **Close this approach** — forcing slice diversity hurts rather than helps
2. If revisiting, try much smaller weight (0.001 or 0.0005) or apply only during early epochs (warmup-then-off)
3. Consider the inverse: a **slice focus loss** that encourages more specialization on surface nodes specifically

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2220 because All 4 metrics regressed significantly (p_in +10.1%, p_tan +5.2%, p_oodc +7.6%, p_re +7.6%). Your analysis is correct and insightful: slice collapse is likely *beneficial* — the model concentrates multiple slices on critical regions (boundary layer, wake, pressure peaks) because those regions need more representational capacity. Forcing diversity redistributes compute to freestream regions that are already well-predicted. The gradient interference between diversity penalty and task loss further hampers training. Slice attention diversity regularization direction exhausted..
